### ISSUE TITLE
messed video quality when recasting (lesson_5_a.py / lesson_5_b.py)

### ISSUE THREAD

**Issue by wkelongws** [elsampsa/valkka-examples#2]
I have an IP camera which outputs 1080x1920 @ 30 fps and a screen shot of the stream played using ffplay is shown below:

![5](https://user-images.githubusercontent.com/14045078/56913201-de941a80-6a65-11e9-9dd6-490ccfeb46aa.png)

After recasting it using the script lesson_5_a.py / lesson_5_b.py, below the screen shot of the recasted stream played using ffplay:

![4](https://user-images.githubusercontent.com/14045078/56913274-0f744f80-6a66-11e9-81e9-ae6230f0e1be.png)

Strong artifacts as observed (in both lesson_5_a.py and lesson_5_b.py).
And below is the message thrown out by ffplay:
```
Input #0, rtsp, from 'rtsp://127.0.0.1:8554/stream1':    0B f=0/0   
  Metadata:
    title           : Session streamed by Valkka
    comment         : stream1
  Duration: N/A, start: 0.040000, bitrate: N/A
    Stream #0:0: Video: h264 (High), yuv420p(tv, bt709, progressive), 1920x1080 [SAR 1:1 DAR 16:9], 29.97 fps, 25 tbr, 90k tbn, 59.94 tbc
[h264 @ 0x7fc650ebb780] error while decoding MB 72 39, bytestream -14
[h264 @ 0x7fc650ebb780] concealing 3457 DC, 3457 AC, 3457 MV errors in P frame
[h264 @ 0x7fc6500a04a0] error while decoding MB 80 37, bytestream -10
[h264 @ 0x7fc6500a04a0] concealing 3689 DC, 3689 AC, 3689 MV errors in P frame
[h264 @ 0x7fc650aaaea0] error while decoding MB 91 35, bytestream -12
[h264 @ 0x7fc650aaaea0] concealing 3918 DC, 3918 AC, 3918 MV errors in P frame
[h264 @ 0x7fc650ebb780] error while decoding MB 30 33, bytestream -12
[h264 @ 0x7fc650ebb780] concealing 4219 DC, 4219 AC, 4219 MV errors in P frame
[h264 @ 0x7fc6500a04a0] error while decoding MB 32 26, bytestream -6
[h264 @ 0x7fc6500a04a0] concealing 5057 DC, 5057 AC, 5057 MV errors in P frame
[h264 @ 0x7fc650aaaea0] error while decoding MB 113 23, bytestream -5
[h264 @ 0x7fc650aaaea0] concealing 5336 DC, 5336 AC, 5336 MV errors in P frame
[h264 @ 0x7fc650ebb780] error while decoding MB 11 23, bytestream -6
[h264 @ 0x7fc650ebb780] concealing 5438 DC, 5438 AC, 5438 MV errors in P frame
[h264 @ 0x7fc6500a04a0] error while decoding MB 5 22, bytestream -7 
[h264 @ 0x7fc6500a04a0] concealing 5564 DC, 5564 AC, 5564 MV errors in P frame
[h264 @ 0x7fc650aaaea0] error while decoding MB 28 22, bytestream -8
...
```

Is there a clue how should I debug the issue? Thanks.

**Comment by elsampsa** [comment]
What's your bitrate?

**Comment by wkelongws** [comment]
about 7 Mb from VLC:
![7](https://user-images.githubusercontent.com/14045078/56914185-4ba8af80-6a68-11e9-99d5-313091f19de1.png)

**Comment by elsampsa** [comment]
Do you get the same artifacts if you configure your camera to a lower bitrate (say, ~ 2000 kbps) ?

**Comment by wkelongws** [comment]
It helps when lower the bit rate but there are still some artifacts there, see the screen shots below:

codec of the original stream from vlc
![7](https://user-images.githubusercontent.com/14045078/56915276-f15d1e00-6a6a-11e9-97c7-6766f277af90.png)

screenshot of the recasted stream in ffplay:
![8](https://user-images.githubusercontent.com/14045078/56915311-06d24800-6a6b-11e9-910d-e607da67e312.png)

**Comment by elsampsa** [comment]
ping: @eranpet : can you take a look at this?

**Comment by elsampsa** [comment]
Ah.. now I got it.  At least for lesson 5a.
- Is the camera coming over wifi? .. lesson 5a is all about multicast, which is notorious when transmitted over wifi
- Once you lowered the bitrate to ~2000 kbps, does lesson 5b (regular unicast) work without artifacts?

**Comment by wkelongws** [comment]
Actually the camera is coming over ethernet and all the screen captures I used here are coming from lesson 5b (rtsp server).

**Comment by elsampsa** [comment]
Hi,

@eranpet has succesfully fixed these issue in libValkka version 0.11.0

Ubuntu's package system is still churning through the compilation process ..

https://launchpad.net/~sampsa-riikonen/+archive/ubuntu/valkka/+packages

.. but after its finished, sudo apt-get upgrade your libValkka version.  Remember also to git pull valkka-examples again

Once you have confirmed that the bug is gone, please close this ticket.

And one thing, when re-transmitting, stream remember to use ```setLiveOutPacketBuffermaxSize``` (see the tutorial).

As per our private discussions, now it is also possible to use multiple cores to decode a single stream:

https://elsampsa.github.io/valkka-examples/_build/html/decoding.html

**Comment by wkelongws** [comment]
Thanks @elsampsa @eranpet. With the 0.11.0 version by increasing the setLiveOutPacketBuffermaxSize() I was able to recast the video smoothly with no artifacts in lesson_5_b.py. 
(for the 7Mb/s stream, the "default" setLiveOutPacketBuffermaxSize(95000) still creates artifacts, by just simply increase it to setLiveOutPacketBuffermaxSize(950000) the issue is gone.)
